captum/optim/_core/objectives.py

```diff
@@ -1,258 +1,349 @@
+        activations = targets_to_values[target]
+        x_diff = activations[..., 1:, :] - activations[..., :-1, :]
+        y_diff = activations[..., :, 1:] - activations[..., :, :-1]
+        return torch.sum(torch.abs(x_diff)) + torch.sum(torch.abs(y_diff))
+
+    return loss_function
+
+
+def l1(target: nn.Module, constant: float = 0) -> LossFunction:
+    """
+    L1 norm of the target layer, generally used as a penalty.
+    """
+
+    def loss_function(targets_to_values: ModuleOutputMapping):
+        activations = targets_to_values[target]
+        return torch.abs(activations - constant).sum()
+
+    return loss_function
+
+
+def l2(target: nn.Module, constant: float = 0, epsilon: float = 1e-6) -> LossFunction:
+    """
+    L2 norm of the target layer, generally used as a penalty.
+    """
+
+    def loss_function(targets_to_values: ModuleOutputMapping):
+        activations = targets_to_values[target]
+        activations = (activations - constant).sum()
+        return torch.sqrt(epsilon + activations)
+
+    return loss_function
+
+
+def diversity(target: nn.Module) -> LossFunction:
+    """
+    Use a cosine similarity penalty to extract features from a polysemantic neuron.
+    Olah, Mordvintsev & Schubert, 2017.
+    https://distill.pub/2017/feature-visualization/#diversity
+    """
+
+    def loss_function(targets_to_values: ModuleOutputMapping):
+        activations = targets_to_values[target]
+        return -sum(
+            [
+                sum(
+                    [
+                        (
+                            torch.cosine_similarity(
+                                activations[j].view(1, -1), activations[i].view(1, -1)
+                            )
+                        ).sum()
+                        for i in range(activations.size(0))
+                        if i != j
+                    ]
+                )
+                for j in range(activations.size(0))
+            ]
+        ) / activations.size(0)
+
+    return loss_function
+
+
+def single_target_objective(
+    target: nn.Module, loss_function: SingleTargetLossFunction
+) -> LossFunction:
+    def inner(targets_to_values: ModuleOutputMapping):
+        value = targets_to_values[target]
+        return loss_function(value)
+
+    return inner
+
+
+class SingleTargetObjective(Objective):
+    def __init__(
+        self,
+        model: nn.Module,
+        target: nn.Module,
+        loss_function: Callable[[torch.Tensor], torch.Tensor],
+    ):
+        super(SingleTargetObjective, self).__init__(model=model, targets=[target])
+        self.loss_function = loss_function
+
+    def loss(self, targets_to_values):
+        assert len(self.targets) == 1
+        target = self.targets[0]
+        target_value = targets_to_values[target]
+        loss_value = self.loss_function(target_value)
+        self.history.append(loss_value.sum().cpu().detach().numpy().squeeze().item())
+        return loss_value
+
+
+class MultiObjective(Objective):

```

Comment:
nit: Do we want to call this: `MultiTargetObjective` to be consistent of `SingleTargetObjective ` ?

Reply:
Yes we probably should call it `MultiTargetObjective` instead of `MultiObjective`. I've changed the name.